src/textual_serve/download_manager.py

```diff
@@ -0,0 +1,148 @@
+    mime_type: str
+    encoding: str | None = None
+    incoming_chunks: asyncio.Queue[bytes | None] = field(default_factory=asyncio.Queue)
+
+
+class DownloadManager:
+    """Class which manages downloads for the server.
+
+    Serves as the link between the web server and app processes during downloads.
+
+    A single server has a single download manager, which manages all downloads for all
+    running app processes.
+    """
+
+    def __init__(self):
+        self._active_downloads_lock = asyncio.Lock()
+        self._active_downloads: dict[str, Download] = {}
+        """A dictionary of active downloads.
+
+        When a delivery key is received in a meta packet, it is added to this set.
+        When the user hits the "/download/{key}" endpoint, we ensure the key is in
+        this set and start the download by requesting chunks from the app process.
+
+        When the download is complete, the app process sends a "deliver_file_end"
+        meta packet, and we remove the key from this set.
+        """
+
+    async def create_download(
+        self,
+        *,
+        app_service: "AppService",
+        delivery_key: str,
+        file_name: str,
+        open_method: str,
+        mime_type: str,
+        encoding: str | None = None,
+    ) -> None:
+        """Prepare for a new download.
+
+        Args:
+            app_service: The app service to start the download for.
+            delivery_key: The delivery key to start the download for.
+            file_name: The name of the file to download.
+            open_method: The method to open the file with.
+        """
+        async with self._active_downloads_lock:
+            self._active_downloads[delivery_key] = Download(
+                app_service,
+                delivery_key,
+                file_name,
+                open_method,
+                mime_type,
+                encoding,
+            )
+
+    async def download(self, delivery_key: str) -> AsyncGenerator[bytes, None]:
+        """Download a file from the given app service.
+
+        Args:
+            app_service: The app service to download from.
+            delivery_key: The delivery key to download.
+        """
+
+        app_service = await self._get_app_service(delivery_key)
+        download = self._active_downloads[delivery_key]
+        incoming_chunks = download.incoming_chunks
+
+        while True:
+            # Request a chunk from the app service.
+            await app_service.send_meta(
+                {
+                    "type": "deliver_chunk_request",
+                    "key": delivery_key,
+                    "size": DOWNLOAD_CHUNK_SIZE,
+                }
+            )
+
+            chunk = await incoming_chunks.get()
+            if not chunk:
+                # Empty chunk - the app process has finished sending the file.
+                incoming_chunks.task_done()
+                with suppress(asyncio.TimeoutError):
+                    await asyncio.wait_for(
```

Comment:
Do we need this join? If we receive an empty chunk doesn't that mean that everything has been processed?

Reply:
Yeah you're right, this is a remnant of a previous approach where I was using a separate "meta" message to signal the end, before switching to just sending the empty chunk.